crypto/batchverifier.go

```diff
@@ -137,50 +149,29 @@ func (b *BatchVerifier) VerifyWithFeedback() (failed []bool, err error) {
 // batchVerificationImpl invokes the ed25519 batch verification algorithm.
 // it returns true if all the signatures were authentically signed by the owners
 // otherwise, returns false, and sets the indexes of the failed sigs in failed
-func batchVerificationImpl(messages [][]byte, publicKeys []SignatureVerifier, signatures []Signature) (allSigsValid bool, failed []bool) {
-
-	numberOfSignatures := len(messages)
-
-	messagesAllocation := C.malloc(C.size_t(C.sizeofPtr * numberOfSignatures))
-	messagesLenAllocation := C.malloc(C.size_t(C.sizeofULongLong * numberOfSignatures))
-	publicKeysAllocation := C.malloc(C.size_t(C.sizeofPtr * numberOfSignatures))
-	signaturesAllocation := C.malloc(C.size_t(C.sizeofPtr * numberOfSignatures))
-	valid := C.malloc(C.size_t(C.sizeof_int * numberOfSignatures))
-
-	defer func() {
-		// release staging memory
-		C.free(messagesAllocation)
-		C.free(messagesLenAllocation)
-		C.free(publicKeysAllocation)
-		C.free(signaturesAllocation)
-		C.free(valid)
-	}()
-
-	// load all the data pointers into the array pointers.
-	for i := 0; i < numberOfSignatures; i++ {
-		*(*uintptr)(unsafe.Pointer(uintptr(messagesAllocation) + uintptr(i*C.sizeofPtr))) = uintptr(unsafe.Pointer(&messages[i][0]))
-		*(*C.ulonglong)(unsafe.Pointer(uintptr(messagesLenAllocation) + uintptr(i*C.sizeofULongLong))) = C.ulonglong(len(messages[i]))
-		*(*uintptr)(unsafe.Pointer(uintptr(publicKeysAllocation) + uintptr(i*C.sizeofPtr))) = uintptr(unsafe.Pointer(&publicKeys[i][0]))
-		*(*uintptr)(unsafe.Pointer(uintptr(signaturesAllocation) + uintptr(i*C.sizeofPtr))) = uintptr(unsafe.Pointer(&signatures[i][0]))
-	}
+func batchVerificationImpl(messages []byte, msgLengths []uint64, publicKeys []SignatureVerifier, signatures []Signature) (allSigsValid bool, failed []bool) {
 
-	// call the batch verifier
-	allValid := C.crypto_sign_ed25519_open_batch(
-		(**C.uchar)(unsafe.Pointer(messagesAllocation)),
-		(*C.ulonglong)(unsafe.Pointer(messagesLenAllocation)),
-		(**C.uchar)(unsafe.Pointer(publicKeysAllocation)),
-		(**C.uchar)(unsafe.Pointer(signaturesAllocation)),
-		C.size_t(len(messages)),
-		(*C.int)(unsafe.Pointer(valid)))
+	numberOfSignatures := len(msgLengths)
+	valid := make([]C.int, numberOfSignatures)
 
+	// call the batch verifier
+	allValid := C.ed25519_batch_wrapper(
+		(*C.uchar)(&messages[0]),
+		(*C.ulonglong)(&msgLengths[0]),
+		(*C.uchar)(&(publicKeys[0][0])),
+		(*C.uchar)(&(signatures[0][0])),
```

Comment:
Is it acceptable to do this?
```suggestion
		(*C.uchar)(&(publicKeys[0])),
		(*C.uchar)(&(signatures[0])),
```
To me that would be clearer, that we're taking the address of the first thing in a linear array. The second index doesn't matter because it's indexing an array rather than slice.

Reply:
I'm claiming that `[][64]byte` _is_ a flat `[]byte`.  That is, the `[]Signature` slice has an underlying array in which "Signatures" (which are really `[64]byte` arrays) are laid out one after another. So if, for example, you have 3 elements in your `signatures` slice, the address of the first byte of the first signature is really the address of at least 192 consecutive bytes containing 3 `[64]byte` arrays. 

This wouldn't be true if the signatures were `[]byte` (a slice), but they are `[64]byte` (an array).

I'll poke around to see if this is correct.

Reply:
For example, this prints numbers that are 32 apart.
```
	if len(signatures) > 1 {
		fmt.Printf("%d %d\n", &(publicKeys[0][0]), &(publicKeys[1][0]))
		panic(1)
	}
```

Reply:
I think the guarantee is that the backing array is one big piece of memory, but agree there's no guarantee that each element touches the next (there could be padding).
At any rate, I misread the later code, I thought it was copying *again*. But it isn't, so I'm happy.